pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java

```diff
@@ -176,49 +195,171 @@ public void markDataResourceOnline(String tableName, ExternalView externalView,
     if (externalView == null) {
       return;
     }
+
+    // Check if there are relevant changes to the state or if we can skip this update
+    if (!isRoutingTableRebuildRequired(tableName, externalView, instanceConfigList)) {
+      return;
+    }
+
+    // Build the routing table
+    buildRoutingTable(tableName, externalView, instanceConfigList);
+  }
+
+  private boolean isRoutingTableRebuildRequired(String tableName, ExternalView externalView,
+      List<InstanceConfig> instanceConfigs) {
+    // In unit tests, always rebuild the routing table
```

Comment:
It was needed when `markDataResourceOnline` checked if there was a rebuild needed, since not all unit tests have a valid HelixManager and/or initialize everything. Originally, everything went through `markDataResourceOnline`, including external view changes, and as such, it needed a rebuild check. Now that `markDataResourceOnline` is unconditional and only depends on the external view and instance configs passed, this workaround isn't required.